File: solutions/typescript/105-course-schedule-ii.ts
/**
 * Problem 105: Course Schedule II (LeetCode 210)
 * Difficulty: Med
 * Language: TypeScript
 */
function findOrder(n: number, prereqs: number[][]): number[] {
  const graph = prereqs.reduce<Record<number, number[]>>(
    (g, [a, b]) => ({ ...g, [b]: [...(g[b] || []), a] }), {}
  );
  type S = { path: Set<number>; done: Set<number>; order: number[]; cycle: boolean };
  const dfs = (v: number, s: S): S => {
    if (s.cycle || s.done.has(v)) return s;
    if (s.path.has(v)) return { ...s, cycle: true };
    const s1 = { ...s, path: new Set([...s.path, v]) };
    const s2 = (graph[v] || []).reduce((st: S, nb: number) => dfs(nb, st), s1);
    return { ...s2, done: new Set([...s2.done, v]), order: [...s2.order, v] };
  };
  const init: S = { path: new Set(), done: new Set(), order: [], cycle: false };
  const result = Array.from({ length: n }, (_, i) => i).reduce((s, i) => dfs(i, s), init);
  return result.cycle ? [] : result.order;
}

File: solutions/elixir/105-course-schedule-ii.ex
# Problem 105: Course Schedule II (LeetCode 210)
# Difficulty: Med
# Language: Elixir
# 
defmodule CourseOrder do
  def find_order(n, prereqs) do
    graph = Enum.reduce(prereqs, %{}, fn [a, b], g ->
      Map.update(g, b, [a], &[a | &1]) end)
    init = %{path: MapSet.new(), done: MapSet.new(), order: [], cycle: false}
    result = Enum.reduce(0..(n-1), init, &dfs(&1, graph, &2))
    if result.cycle, do: [], else: result.order
  end
  defp dfs(v, graph, s) do
    cond do
      s.cycle or MapSet.member?(s.done, v) -> s
      MapSet.member?(s.path, v) -> %{s | cycle: true}
      true ->
        s1 = %{s | path: MapSet.put(s.path, v)}
        s2 = Enum.reduce(Map.get(graph, v, []), s1, &dfs(&1, graph, &2))
        %{s2 | done: MapSet.put(s2.done, v), order: s2.order ++ [v]}
    end
  end
end
